mlir/include/mlir/IR/CommonTypeConstraints.td

```diff
@@ -481,10 +500,39 @@ class IsScalableVectorOfLengthPred<list<int> allowedLengths> :
                            == }]
                          # allowedlength>)>]>;
 
+// Normalizes an index so it can be bounds checked.
+// Negative values are mapped to their absolute value.
+//  - These are used to index in reverse (i.e. index -1 would be the last element)
+// Positive values are mapped to their value + 1.
+//  - This results the same range of values as the negative indices
+// This allows bounds checking to be: len(list) >= NormalizeIndex<idx>.ret.
+class NormalizeIndex<int value> {
+  int ret = !if(!lt(value, 0), !sub(0, value), !add(value, 1));
+}
```

Comment:
It says it in the comment `"This allows bounds checking to be: len(list) >= NormalizeIndex<idx>.ret"`. Otherwise, you need to do `idx > 0 && idx < len(list) || idx < 0 && -idx <= len(list)` which is not easy to write in tablegen.

Reply:
(writing the logic with !if/!sub!/!add, is much easier than a chain of CPreds)

Reply:
Normalized here means the indices in both directions have the same value. 
I.e. The first element going forwards is index 2, going backwards it's index -3.
So you normalize it so they're both 3.

Reply:
(this was previously just `abs()`, but it was changed to this to allow indexing from 0 when going forwards).